### PR TITLE
ncm-ceph: various additions

### PR DESCRIPTION
**Pull request by kwaegema**
- add `max_add_osd_failures_per_host` option to be flexible with osd failures
- change schema to be able to use per host osd templates that are included on the deployhost.
- change injection to a warning: injecting on a host base does not work (anymore)
- remove empty crushmap buckets when labeling

### CODE REVIEW

**Comment by kwaegema** [comment]
fixes #419

File: ncm-ceph/src/main/pan/components/ceph/schema.pan [hunk 1]
```diff
@@ -318,14 +321,24 @@ type ceph_cluster = {
     'crushmap'                  ? ceph_crushmap
 };
 
+@documentation{
+Decentralized config feature:
+For use with dedicated pan code that builds the cluster info from remote templates..
```
Comment (stdweird) [inline ncm-ceph/src/main/pan/components/ceph/schema.pan]: `..` -> `.`

File: ncm-ceph/src/main/pan/components/ceph/schema.pan [hunk 2]
```diff
@@ -318,14 +321,24 @@ type ceph_cluster = {
...
+type ceph_localdaemons = {
+    'osds'  : ceph_osd {}
+};
+
 @documentation{ ceph clusters }
 type ${project.artifactId}_component = {
     include structure_component
-    'clusters'         : ceph_cluster {}
+    'clusters'         ? ceph_cluster {}
+    'localdaemons'     ? ceph_localdaemons #validation, but not used in component code
```
Comment (stdweird) [inline ncm-ceph/src/main/pan/components/ceph/schema.pan]: space after `#`

File: ncm-ceph/src/main/perl/Ceph/config.pm [hunk 3]
```diff
@@ -89,10 +89,12 @@ sub inject_realtime {
         if (!($param ~~ @NONINJECT)) { # Requires Perl > 5.10 !
             @cmd = ('tell',"*.$host",'injectargs','--');
             my $keyvalue = "--$param=$changes->{$param}";
-            $self->info("injecting $keyvalue realtime on $host");
-            $self->run_ceph_command([@cmd, $keyvalue]) or return 0;
+            #$self->info("injecting $keyvalue realtime on $host"); # this does not work anymore, see mailinglist
```
Comment (stdweird) [inline ncm-ceph/src/main/perl/Ceph/config.pm]: remove if unused

**Comment by stdweird** [comment]
@kwaegema minor remarks, LGTM otherwise

**Comment by kwaegema** [comment]
Does this needs additional review?

File: ncm-ceph/src/main/pan/components/ceph/schema.pan [hunk 4]
```diff
@@ -13,23 +13,26 @@ include { 'quattor/schema' };
 }
 function valid_osd_names = {
     names = list();
-    clusters = ARGV[0]['clusters'];
-    foreach (name;cluster;clusters) {
-        append(names, name);
-    };
+    if(exists(ARGV[0]['clusters'])) {
```
Comment (stdweird) [inline ncm-ceph/src/main/pan/components/ceph/schema.pan]: avoid the extra indent, do a 
```pan
if(! exists(ARGV[0]['clusters'])) { return true; };
```
(also, set it before the `names = list();`

File: ncm-ceph/src/main/perl/Ceph/daemon.pm [hunk 5]
```diff
@@ -208,6 +208,11 @@ sub check_empty {
             return 0;
         }
     } else {
+        my $checkmntcmd = ['/usr/bin/findmnt', $loc ];
```
Comment (stdweird) [inline ncm-ceph/src/main/perl/Ceph/daemon.pm]: what package provides this? does it require a dependency?

File: ncm-ceph/src/main/perl/Ceph/crushmap.pm [hunk 6]
```diff
@@ -119,6 +119,10 @@ sub crush_merge {
                     my $osds = $osdhosts->{$name}->{osds};
                     $bucket->{buckets} = [];
                     foreach my $osd (sort(keys %{$osds})){ 
+                        if ($osds->{$osd}->{crush_ignore}){
+                            $self->warn("Osd $osd on $name was set to ignore");
```
Comment (stdweird) [inline ncm-ceph/src/main/perl/Ceph/crushmap.pm]: `Osd` -> `OSD`

File: ncm-ceph/src/main/perl/ceph.pm [hunk 7]
```diff
@@ -252,7 +252,8 @@ sub Configure {
             hostname => $hostname,
             is_deploy => $is_deploy,
             cephusr => $cephusr,
-            key_accept => $t->{key_accept} 
+            key_accept => $t->{key_accept}, 
+            max_add_osd_failures_per_host => $t->{max_add_osd_failures_per_host}
```
Comment (stdweird) [inline ncm-ceph/src/main/perl/ceph.pm]: end with `,`

File: ncm-ceph/src/main/perl/Ceph/daemon.pm (hunk 5 above)
Comment (kwaegema) [inline ncm-ceph/src/main/perl/Ceph/daemon.pm]: @stdweird util-linux. I thought it was by default included in OS, but apparently we include it in a template for all our servers:)

Comment (kwaegema) [inline ncm-ceph/src/main/perl/Ceph/daemon.pm]: afaik, it is a standard package. But I've just seen that on el6 systems it is on a different path, /bin/findmnt..

**Comment by kwaegema** [comment]
remarks processed

**Comment by jrha** [comment]
LGTM

**Comment by stdweird** [comment]
LGTM

**Pull request merged**